packages/context/src/context.ts

```diff
@@ -96,10 +129,75 @@ export class Context {
   unbind(key: BindingAddress): boolean {
     key = BindingKey.validate(key);
     const binding = this.registry.get(key);
+    // If not found, return `false`
     if (binding == null) return false;
     if (binding && binding.isLocked)
       throw new Error(`Cannot unbind key "${key}" of a locked binding`);
-    return this.registry.delete(key);
+    this.registry.delete(key);
+    this.notifyListeners('unbind', binding);
+    return true;
+  }
+
+  /**
+   * Add the context listener as an event listener to the context chain,
+   * including its ancestors
+   * @param listener Context listener
+   */
+  subscribe(listener: ContextEventListener): Subscription {
+    let ctx: Context | undefined = this;
+    while (ctx != null) {
+      ctx.listeners.add(listener);
+      ctx = ctx._parent;
+    }
+    return new ContextSubscription(this, listener);
+  }
+
+  /**
+   * Remove the context listener  from the context chain
+   * @param listener Context listener
+   */
+  unsubscribe(listener: ContextEventListener) {
+    let ctx: Context | undefined = this;
+    while (ctx != null) {
+      ctx.listeners.delete(listener);
+      ctx = ctx._parent;
+    }
+  }
+
+  /**
+   * Check if a listener is subscribed to this context
+   * @param listener Context listener
+   */
+  isSubscribed(listener: ContextEventListener) {
+    return this.listeners.has(listener);
+  }
+
+  /**
+   * Publish an event to the registered listeners. Please note the
+   * notification happens using `process.nextTick` so that we allow fluent APIs
+   * such as `ctx.bind('key').to(...).tag(...);` and give listeners the fully
+   * populated binding
+   *
+   * @param event Event names: `bind` or `unbind`
+   * @param binding Binding bound or unbound
+   */
+  protected notifyListeners(
+    event: ContextEventType,
+    binding: Readonly<Binding<unknown>>,
+  ) {
+    // Notify listeners in the next tick
+    process.nextTick(async () => {
```

Comment:
This isn't going to work well. The promise returned by the async function is ignored by `process.nextTick`. When an error happens, a warning "unhandled promise rejection" is triggered. In the future versions of Node, this can crash the app the same way as "uncaught error" does already.

I'd like us to consider two aspects here.

1. Do we want `notifyListeners` to wait until all listeners processed the notification? In the proposal here, we treat notifications as "fire and forget" and run them kind of in background.

2. If we decide for "fire and forget" mode, how should we handle errors reported by listeners? IMO, application developers should have a way how to catch such errors and handle them in their own way.

I am imagining that the currently proposed solution will make it more difficult to write a reliable test for ContextView.

```ts
const context = new Context();
const view = new ContextView(view);

context.bind('foo').to('bar');
// ^^ triggers notification at some point in the future

const values = await view.resolve();
// ^^ how to ensure this is run only after the view was notified
// about the new binding?

expect(values).to.containEqual('bar');
```

I wanted to propose to wait until all listeners processed the notification:
- Error handling is much easier, we can propagate any errors thrown by listeners up the call chain (down the call stack).
- We avoid subtle errors that can happen when the notification of a listener is executed later than code relying on the outcome of that listener.

I was thinking about the following, just to realize that it won't work either:

```ts
const context = new Context();
const view = new ContextView(view);

await context.bind('foo').to('bar');
// ^^ uh oh, what are we trying to await?
// The outcome of `Binding.prototype.to()`?!

const values = await view.resolve();
expect(values).to.containEqual('bar');
```

Also changing the act of creating a new binding from sync to async is a big change that's possibly backwards incompatible.

My conclusion is that we should keep "fire and forget" approach but need to rework the way how these delayed notifications work. Instead of creating a new `process.nextTick` callback for each new binding, we should maintain a queue of pending notifications and a single `process.nextTick` callback to process the entire queue. 

```ts
/*** schedule a new notification ***/
this._pendingNotifications.push({listener, event, args: [binding]});
if (!this._queueHandlerScheduled) {
  process.nextTick(this._processQueueCallback, this);
  this._queueHandlerScheduled = true;
}

/*** process notifications ***/
_processQueueCallback() {
  this._queueHandlerScheduled = false;
  this._processQueueAsync.then(
    success => {
      // let Context consumers know that we are done, e.g.
      this.emit('NotificationsProcessed');
    },
    error => {
      // signal the error to Context consumers, e.g.
      this.emit('error', error);
    }
  );
}

async _processQueueAsync() {
  const notifications = this. _pendingNotifications;
  this._pendingNotifications = [];

  return Promise.all(notifications.map(
    n => n.listener.listen(n.event, ...n.args)
  ));
}

/*** usage in the tests ***/
context.bind('foo').to('bar');
await pEvent(context, 'NotificationsProcessed');
```

@raymondfeng @hacksparrow @jannyHou what do you think?

Regardless of the solution we decide to implement, I think it's very important to document the reasons why we decided for a particular design & implementation, the alternatives we have rejected and why they were rejected.

I am thinking about _Lightweight Architecture Decision Records_ as recommended by [ThoughWork's technology radar](https://www.thoughtworks.com/radar/techniques/lightweight-architecture-decision-records):

> Much documentation can be replaced with highly readable code and tests. In a world of evolutionary architecture, however, it's important to record certain design decisions for the benefit of future team members as well as for external oversight. Lightweight Architecture Decision Records is a technique for capturing important architectural decisions along with their context and consequences. We recommend storing these details in source control, instead of a wiki or website, as then they can provide a record that remains in sync with the code itself. For most projects, we see no reason why you wouldn't want to use this technique.

If it makes it easier for us, then I am also ok with keeping the documentation inside the source code as a (somewhat longer) code comment.

Reply:
Based on my last comment below, it may be better to process notifications in series.

```diff
 async _processQueueAsync() {
   const notifications = this. _pendingNotifications;
   this._pendingNotifications = [];

-  return Promise.all(notifications.map(
-    n => n.listener.listen(n.event, ...n.args)
-  ));
+  for (const n of notifications) {
+    await n.listener.listen(n.event, ...n.args)
+  }
 }
```

Reply:
> As an app developer, I would want to register my error listener on the application context and receive errors from child contexts too, e.g. from per-request contexts.

Yes. We should also we able to find out which context triggered the error. 

> Do we really need to support async listeners? 

I think we should. We should not enforce blocking code in our framework. Developers should have a choice to use sync or async version of an API.

Reply:
> I'd prefer to start with implementation that's simpler - i.e. sync listeners only, and wait for user feedback & demand for async listeners before we add all the complexity.

On the other hand, because we want to run even synchronous listeners later, on a clean stack started by `process.nextTick()`, we have to address all issues caused by asynchronicity anyways. At that point the matter of adding support for async listeners is trivial and makes sense to get done.